File: Server/Gatekeeper/src/main/java/org/elkoserver/server/gatekeeper/GatekeeperBoot.java
package org.elkoserver.server.gatekeeper;

import java.time.Clock;
import java.util.List;
import java.util.Set;
import org.elkoserver.foundation.boot.Bootable;
import org.elkoserver.foundation.net.MessageHandlerFactory;
import org.elkoserver.foundation.properties.ElkoProperties;
import org.elkoserver.foundation.server.Server;
import org.elkoserver.foundation.server.ServiceFactory;
import org.elkoserver.foundation.server.metadata.AuthDesc;
import org.elkoserver.foundation.timer.Timer;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

/**
 * The Elko boot class for the Gatekeeper.  The Gatekeeper is a server that
 * provides login reservation and authentication services for other Elko
 * servers such as the Director.
 */
public class GatekeeperBoot implements Bootable {
    private TraceFactory traceFactory;
    private Trace tr;
    private Timer timer;
    private Gatekeeper myGatekeeper;

    /** How long user has before being kicked off, in milliseconds. */
    private int myActionTimeout;

    /** Default action timeout, in seconds. */
    private static final int DEFAULT_ACTION_TIMEOUT = 15;

    public void boot(ElkoProperties props, TraceFactory traceFactory, Clock clock) {
        this.traceFactory = traceFactory;
        tr = traceFactory.trace("gate");
        timer = new Timer(traceFactory, clock);
        Server server = new Server(props, "gatekeeper", tr, timer, clock, traceFactory);

        myGatekeeper = new Gatekeeper(server, tr, timer, traceFactory);
        myActionTimeout = 1000 *
            props.intProperty("conf.gatekeeper.actiontimeout",
                              DEFAULT_ACTION_TIMEOUT);

        if (server.startListeners("conf.listen",
                                  new GatekeeperServiceFactory()) == 0) {
            tr.errori("no listeners specified");
        }
    }

    /**
     * Service factory for the Gatekeeper.
     *
     * The Gatekeeper offers two kinds of service connections:
     *
     *    gatekeeper/user     - for clients seeking services
     *    gatekeeper/admin    - for system administrators
     */
    private class GatekeeperServiceFactory implements ServiceFactory {

        public MessageHandlerFactory provideFactory(String label,
                                                    AuthDesc auth,
                                                    Set<String> allow,
                                                    List<String> serviceNames,
                                                    String protocol)
        {
            boolean allowUser = false;
            boolean allowAdmin = false;

            if (allow.contains("any")) {
                allowUser = true;
                allowAdmin = true;
            } 
            if (allow.contains("admin")) {
                allowAdmin = true;
            }
            if (allow.contains("user")) {
                allowUser = true;
            }

            if (allowAdmin) {
                serviceNames.add("gatekeeper-admin");
            }
            if (allowUser) {
                serviceNames.add("gatekeeper-user");
            }
            return new GatekeeperActorFactory(myGatekeeper, auth, allowAdmin,
                                              allowUser, myActionTimeout, tr, timer, traceFactory);
        }
    }
}

File: Server/Gatekeeper/src/main/kotlin/org/elkoserver/server/gatekeeper/GatekeeperBoot.kt
package org.elkoserver.server.gatekeeper

import org.elkoserver.foundation.boot.Bootable
import org.elkoserver.foundation.net.Communication.COMMUNICATION_CATEGORY_TAG
import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.foundation.server.ShutdownWatcher
import org.elkoserver.util.trace.slf4j.Gorgel
import org.ooverkommelig.ConstantDefinition
import org.ooverkommelig.ObjectGraphConfiguration
import org.ooverkommelig.ObjectGraphLogger
import java.time.Clock

/**
 * The Elko boot class for the Gatekeeper.  The Gatekeeper is a server that
 * provides login reservation and authentication services for other Elko
 * servers such as the Director.
 */
@Suppress("unused")
class GatekeeperBoot : Bootable {
    override fun boot(props: ElkoProperties, gorgel: Gorgel, clock: Clock) {
        val myGorgel = gorgel.getChild(GatekeeperBoot::class)
        lateinit var gatekeeperServerGraph: GatekeeperServerOgd.Graph
        val graphClosingShutdownWatcher = ShutdownWatcher { gatekeeperServerGraph.close() }
        gatekeeperServerGraph = GatekeeperServerOgd(object : GatekeeperServerOgd.Provided {
            override fun clock() = ConstantDefinition(clock)
            override fun baseGorgel() = ConstantDefinition(gorgel)
            override fun baseCommGorgel() =
                ConstantDefinition(gorgel.withAdditionalStaticTags(COMMUNICATION_CATEGORY_TAG))

            override fun props() = ConstantDefinition(props)
            override fun externalShutdownWatcher() = ConstantDefinition(graphClosingShutdownWatcher)
        }, ObjectGraphConfiguration(object : ObjectGraphLogger {
            override fun errorDuringCleanUp(sourceObject: Any, operation: String, exception: Exception) {
                myGorgel.error(
                    "Error during cleanup of object graph. Object: $sourceObject, operation: $operation",
                    exception
                )
            }
        })).Graph()
        gatekeeperServerGraph.server()
    }

    companion object {
        /** Default action timeout, in seconds.  */
        internal const val DEFAULT_ACTION_TIMEOUT = 15
    }
}
